alibi_detect/cd/base.py

```diff
@@ -0,0 +1,215 @@
+from alibi_detect.cd.preprocess import preprocess_drift
+from alibi_detect.cd.utils import update_reference
+from alibi_detect.utils.statstest import fdr
+
+logger = logging.getLogger(__name__)
+
+
+class BaseUnivariateDrift(BaseDetector):
+
+    def __init__(self,
+                 p_val: float = .05,
+                 X_ref: Union[np.ndarray, list] = None,
+                 preprocess_X_ref: bool = True,
+                 update_X_ref: Optional[Dict[str, int]] = None,
+                 preprocess_fn: Optional[Callable] = None,
+                 preprocess_kwargs: Optional[dict] = None,
+                 correction: str = 'bonferroni',
+                 n_features: Optional[int] = None,
+                 n_infer: int = 2,
+                 input_shape: Optional[tuple] = None,
+                 data_type: Optional[str] = None
+                 ) -> None:
+        """
+        Generic drift detector components which serves as a base class for methods using
+        univariate tests with multivariate correction.
+
+        Parameters
+        ----------
+        p_val
+            p-value used for significance of the statistical test for each feature. If the FDR correction method
+            is used, this corresponds to the acceptable q-value.
+        X_ref
+            Data used as reference distribution.
+        preprocess_X_ref
+            Whether to already preprocess and store the reference data.
+        update_X_ref
+            Reference data can optionally be updated to the last n instances seen by the detector
+            or via reservoir sampling with size n. For the former, the parameter equals {'last': n} while
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+            Typically a dimensionality reduction technique.
+        preprocess_kwargs
+            Kwargs for `preprocess_fn`.
+        correction
+            Correction type for multivariate data. Either 'bonferroni' or 'fdr' (False Discovery Rate).
+        n_features
+            Number of features used in the statistical test. No need to pass it if no preprocessing takes place.
+            In case of a preprocessing step, this can also be inferred automatically but could be more
+            expensive to compute.
+        n_infer
+            Number of instances used to infer number of features from.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__()
+
+        if p_val is None:
+            logger.warning('No p-value set for the drift threshold. Need to set it to detect data drift.')
+
+        if isinstance(preprocess_kwargs, dict):  # type: ignore
+            if not isinstance(preprocess_fn, Callable):  # type: ignore
```

Comment:
What are the `mypy` errors for these (and the following) annotations?

Reply:
` error: Argument 2 to "isinstance" has incompatible type "_SpecialForm"; expected "Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]"`

Reply:
Upon further investigating, `Callable` is not a valid option to `isinstance` resulting in the `mypy` errors. Solution is to check `callable(preprocess_fn)` which returns a `bool`.

Reply:
Disregard above as the issue lies with mypy and the current check is idiomatic.